src/verbs/queue_pair.rs

```diff
@@ -598,3 +722,156 @@ fn attr_mask_check(
+        self.guard.setup_send();
+        SendHandle { guard: self.guard }
+    }
+
+    pub fn setup_write(self, rkey: u32, remote_addr: u64) -> WriteHandle<'g, G> {
+        self.guard.setup_write(rkey, remote_addr);
+        WriteHandle { guard: self.guard }
+    }
+}
+
+pub struct BasicPostSendGuard<'qp> {
+    qp: NonNull<ibv_qp>,
+    _phantom: PhantomData<&'qp ()>,
+}
+
+impl PostSendGuard for BasicPostSendGuard<'_> {
+    fn construct_wr<'qp>(&'qp mut self, wr_id: u64, wr_flags: WorkRequestFlags) -> WorkRequestHandle<'qp, Self> {
+        todo!()
+    }
+
+    fn post(self) -> Result<(), String> {
+        todo!()
+    }
+}
+
+impl private_traits::PostSendGuard for BasicPostSendGuard<'_> {
+    fn setup_send(&mut self) {
+        todo!()
+    }
+
+    fn setup_write(&mut self, rkey: u32, remote_addr: u64) {
+        todo!()
+    }
+
+    fn setup_inline_data(&mut self, buf: &[u8]) {
+        todo!()
+    }
+
+    fn setup_inline_data_list(&mut self, buf: &[IoSlice<'_>]) {
+        todo!()
+    }
+}
+
+pub struct ExtendedPostSendGuard<'qp> {
+    qp_ex: Option<NonNull<ibv_qp_ex>>,
+    _phantom: PhantomData<&'qp ()>,
+}
+
+impl PostSendGuard for ExtendedPostSendGuard<'_> {
+    fn construct_wr<'qp>(&'qp mut self, wr_id: u64, wr_flags: WorkRequestFlags) -> WorkRequestHandle<'qp, Self> {
+        unsafe {
+            self.qp_ex.unwrap_unchecked().as_mut().wr_id = wr_id;
+            self.qp_ex.unwrap_unchecked().as_mut().wr_flags = wr_flags.bits;
+        }
+        WorkRequestHandle { guard: self }
+    }
+
+    fn post(mut self) -> Result<(), String> {
+        let ret: i32 = unsafe { ibv_wr_complete(self.qp_ex.unwrap_unchecked().as_ptr()) };
+
+        self.qp_ex = None;
+
+        match ret {
+            0 => Ok(()),
+            err => Err(format!("failed to ibv_wr_complete: ret {err}")),
+        }
+    }
+}
+
+impl private_traits::PostSendGuard for ExtendedPostSendGuard<'_> {
+    fn setup_send(&mut self) {
+        unsafe {
+            ibv_wr_send(self.qp_ex.unwrap_unchecked().as_ptr());
+        }
+    }
+
+    fn setup_write(&mut self, rkey: u32, remote_addr: u64) {
+        unsafe {
+            ibv_wr_rdma_write(self.qp_ex.unwrap_unchecked().as_ptr(), rkey, remote_addr);
+        }
+    }
+
+    fn setup_inline_data(&mut self, buf: &[u8]) {
+        unsafe {
+            self.qp_ex.unwrap_unchecked().as_mut().wr_flags |= WorkRequestFlags::Inline.bits;
+            ibv_wr_set_inline_data(self.qp_ex.unwrap_unchecked().as_ptr(), buf.as_ptr() as _, buf.len())
+        }
+    }
+
+    fn setup_inline_data_list(&mut self, buf: &[IoSlice<'_>]) {
+        unsafe {
+            self.qp_ex.unwrap_unchecked().as_mut().wr_flags |= WorkRequestFlags::Inline.bits;
+        }
+        let mut buf_list: Vec<ibv_data_buf> = Vec::new();
+        for slice in buf {
+            buf_list.push(ibv_data_buf {
+                addr: slice.as_ptr() as _,
+                length: slice.len(),
+            })
+        }
```

Comment:
We can impl this in a more efficient & concise way:
```Rust
        let mut buf_list = Vec::with_capacity(buf.len());
        buf_list.extend(buf.iter().map(|x| ibv_data_buf {
            addr: x.as_ptr() as _,
            length: x.len(),
        }));
```

Reply:
What do you think about this
```rust
        let buf_list: Vec<ibv_data_buf> = bufs
            .iter()
            .map(|x| ibv_data_buf {
                addr: x.as_ptr() as _,
                length: x.len(),
            })
            .collect();
```
Now that we would use `iter()` and `map()`, we can just `collect()` them.

Reply:
I'm not sure whether using `collect()` will reserve enough capacity in the first place

Reply:
After taking a look at this thread https://github.com/rust-lang/rust/issues/48994, got to know that `collect()` would take a size_hint into account, though only lower bound is considered, so the Vec actually would be resized in the middle. I would revert this back.